File: Data Structures & Algorithms/implement-prefix-tree/submission-0.py
class CharNode:
    def __init__(self):
        self.next = {}
        self.isEnd = False

class PrefixTree:

    def __init__(self):
        self.head = CharNode()

    def insert(self, word: str) -> None:
        cur = self.head
        for c in word:
            if c not in cur.next:
                cur.next[c] = CharNode()
            cur = cur.next[c]
          
        cur.isEnd = True

    def search(self, word: str) -> bool:
        cur = self.head
        for c in word:
            if c not in cur.next:
                return False
            cur = cur.next[c]
        
        return cur.isEnd
        

    def startsWith(self, prefix: str) -> bool:
        cur = self.head
        for c in prefix:
            if c not in cur.next:
                return False
            cur = cur.next[c]
        
        return True
        
        

File: Data Structures & Algorithms/remove-sub-folders-from-the-filesystem/submission-0.cpp
struct Node {
    Node() : wordEnd(false) {}

    std::unordered_map<char, std::unique_ptr<Node>> next;
    bool wordEnd;
};

class PrefixTree {
public:
    PrefixTree() {
        head = std::make_unique<Node>();
    }
    
    void insert(const string& word) {
        Node* cur = head.get();
        for (const auto c : word) {
            if (cur->next.count(c) == 0) {
                cur->next[c] = std::make_unique<Node>();
            }
            cur = cur->next[c].get();
        }
        cur->wordEnd = true;
    }
    
    bool search(const string& word) {
        Node* cur = head.get();
        for (const auto c : word) {
            if (cur->next.count(c) == 0) {
                return false;
            }
            cur = cur->next[c].get();
        }
        return cur->wordEnd;
    }
    
    bool startsWith(string prefix) {
        Node* cur = head.get();
        for (const auto c : prefix) {
            if (cur->next.count(c) == 0) {
                return false;
            }
            cur = cur->next[c].get();
        }
        return true;
    }

    std::unique_ptr<Node> head;
};

class Solution {
public:
    vector<string> removeSubfolders(vector<string>& folder) {
        PrefixTree t;
        vector<string> ans;
        std::sort(folder.begin(), folder.end());

        for (const auto& f : folder) {
            Node* cur = t.head.get();
            bool isSubFolder = false;
            int idx = 0;
            for (const auto c : f) {
                if (cur->next.count(c) == 0) {
                    cur->next[c] = std::make_unique<Node>();
                }

                cur = cur->next[c].get();

                if (cur->wordEnd && idx + 1 < f.size() && f[idx + 1] == '/') {
                    isSubFolder = true;
                    break;
                }

                ++idx;

            }

            if (isSubFolder) {
                continue;
            }

            cur->wordEnd = true;
            ans.emplace_back(f);
        }

        return ans;
    }
};